File: services/aep-api/internal/spec/descriptor_test.go
// Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
//
// WSO2 LLC. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package spec

import (
	"strings"
	"testing"
)

// The idea is free text a user typed: newlines, straight quotes, apostrophes
// and backslashes all have to survive a write→read cycle intact. This is the
// whole reason the descriptor uses a real TOML encoder instead of hand-rolled
// key writing.
func TestDescriptorRoundTripsAwkwardIdeaText(t *testing.T) {
	t.Parallel()
	idea := "A \"claims\" tracker for Ops.\nDon't lose receipts — path C:\\temp matters.\n\n[not-a-section]\nidea = \"not a key\""
	in := Descriptor{
		APIVersion: DescriptorAPIVersion,
		Name:       "expense-tracker",
		CreatedAt:  "2026-07-29T10:14:00Z",
		Idea:       idea,
	}

	raw, err := MarshalDescriptor(in)
	if err != nil {
		t.Fatalf("MarshalDescriptor: %v", err)
	}
	got, err := ParseDescriptor(raw)
	if err != nil {
		t.Fatalf("ParseDescriptor(%q): %v", raw, err)
	}
	if got.Idea != idea {
		t.Fatalf("idea round-trip:\n got %q\nwant %q\nencoded as:\n%s", got.Idea, idea, raw)
	}
	if got.Name != in.Name || got.APIVersion != in.APIVersion || got.CreatedAt != in.CreatedAt {
		t.Fatalf("identity round-trip = %+v, want %+v", got, in)
	}
}

// A hand-written descriptor (the shape documented for users) parses.
func TestParseDescriptorHandWritten(t *testing.T) {
	t.Parallel()
	const raw = `apiVersion = "agentic-engineer/v1"
name = "expense-tracker"
createdAt = "2026-07-29T10:14:00Z"

idea = """
An expense claim tracker for a 200-person company,
with manager approval and receipt uploads.
"""
`
	got, err := ParseDescriptor([]byte(raw))
	if err != nil {
		t.Fatalf("ParseDescriptor: %v", err)
	}
	if got.Name != "expense-tracker" {
		t.Fatalf("name = %q", got.Name)
	}
	if !strings.HasPrefix(got.Idea, "An expense claim tracker") ||
		!strings.Contains(got.Idea, "receipt uploads.") {
		t.Fatalf("idea = %q", got.Idea)
	}
}

func TestParseDescriptorRejectsGarbage(t *testing.T) {
	t.Parallel()
	if _, err := ParseDescriptor([]byte("this is not = = toml [[[")); err == nil {
		t.Fatal("want a parse error on malformed TOML, got nil")
	}
}

// NewDescriptor stamps the current apiVersion so callers cannot forget it —
// the field is what identifies the file as an Agentic Engineer descriptor.
func TestNewDescriptorStampsAPIVersion(t *testing.T) {
	t.Parallel()
	d := NewDescriptor("expense-tracker", "an expense tracker", "2026-07-29T10:14:00Z")
	if d.APIVersion != DescriptorAPIVersion {
		t.Fatalf("apiVersion = %q, want %q", d.APIVersion, DescriptorAPIVersion)
	}
}

// ideaSteer is the server half that carries the captured idea into a
// requirements-generate turn. Empty/whitespace idea appends NOTHING, so a turn
// without a descriptor is byte-identical to one before this feature.
func TestIdeaSteer(t *testing.T) {
	t.Parallel()
	if got := ideaSteer("   \n  "); got != "" {
		t.Fatalf("blank idea must append nothing, got %q", got)
	}
	got := ideaSteer("  an expense claim tracker  ")
	if !strings.HasPrefix(got, "\n\n") {
		t.Fatalf("steer must start with the paragraph break, got %q", got)
	}
	if !strings.Contains(got, "an expense claim tracker") {
		t.Fatalf("steer must carry the idea verbatim, got %q", got)
	}
	if strings.Contains(got, "  an expense") {
		t.Fatalf("steer must trim the idea, got %q", got)
	}
}

File: playground/test/descriptor.test.ts
/**
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * The project descriptor (`specs/.agentic-engineer.toml`) and the `/start`
 * kickoff it feeds: the playground's half of what aep-api does server-side.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync, mkdirSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { DESCRIPTOR_PATH, descriptorFile, readIdea, writeDescriptor } from "../src/state/descriptor.js";
import { startInstruction } from "../src/engine/compose.js";
import { startInstruction as startInstructionText, ideaSteerPrefix } from "@aep/contracts/prompts";
import { classifyChatInput } from "../src/tui/chat-commands.js";
import { readProjectFiles } from "../src/kit/project-fs.js";
import { renderPart } from "../src/kit/render.js";

function tempProject(): string {
  return mkdtempSync(join(tmpdir(), "aep-desc-test-"));
}

// The idea is free text a user typed — quotes, apostrophes, backslashes and
// newlines must survive the write→read cycle. This is why a real TOML encoder
// is used rather than hand-rolled key writing.
test("descriptor round-trips awkward idea text", () => {
  const dir = tempProject();
  try {
    const idea = 'A "claims" tracker for Ops.\nDon\'t lose receipts — path C:\\temp matters.\n\n[not-a-section]\nidea = "not a key"';
    writeDescriptor(dir, "expense-tracker", idea);
    assert.equal(readIdea(dir), idea);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

test("readIdea returns null when there is no descriptor", () => {
  const dir = tempProject();
  try {
    assert.equal(readIdea(dir), null);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

// A hand-mangled descriptor must not break a run: losing the idea costs one
// question from the start skill, which is cheaper than a crash.
test("readIdea degrades to null on malformed TOML", () => {
  const dir = tempProject();
  try {
    const file = descriptorFile(dir);
    mkdirSync(dirname(file), { recursive: true });
    writeFileSync(file, "this is not = = toml [[[", "utf8");
    assert.equal(readIdea(dir), null);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

// THE load-bearing property: the descriptor is invisible to the agent. The
// snapshot walk strips dot-led segments at every depth, so the model can never
// read the idea for itself — it only ever arrives through the /start expansion.
test("the descriptor never enters a turn snapshot", () => {
  const dir = tempProject();
  try {
    writeDescriptor(dir, "expense-tracker", "an expense claim tracker");
    mkdirSync(join(dir, "specs/requirements"), { recursive: true });
    writeFileSync(join(dir, "specs/requirements/prd.md"), "# Reqs\n", "utf8");

    const files = readProjectFiles(dir);
    assert.ok(files["specs/requirements/prd.md"], "ordinary spec files still ride");
    assert.equal(files[DESCRIPTOR_PATH], undefined, "the descriptor must never reach the agent");
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

// --- /start classification + expansion ---------------------------------------

test("/start classifies as the kickoff, with or without an inline idea", () => {
  assert.deepEqual(classifyChatInput("/start"), { kind: "start" });
  assert.deepEqual(classifyChatInput("/start a rota planner"), { kind: "start", inlineIdea: "a rota planner" });
  // Trailing whitespace only is still a bare kickoff, not an empty idea.
  assert.deepEqual(classifyChatInput("/start   "), { kind: "start" });
});

// The grammar is narrow so real chat is never eaten.
test("prose mentioning /start is an ordinary turn", () => {
  const intent = classifyChatInput("where do I /start with the design?");
  assert.equal(intent.kind, "turn");
});

// A `/start` interview ends its turn on a question card, whose tool result is a
// placeholder carrying `status` but no `ok`. Rendering that as an error was
// telling every new user their first command failed.
test("a pending question card renders as awaiting, not as an error", () => {
  const written: string[] = [];
  const original = process.stdout.write.bind(process.stdout);
  (process.stdout as { write: unknown }).write = (chunk: string) => {
    written.push(String(chunk));
    return true;
  };
  try {
    renderPart({
      type: "tool-result",
      toolCallId: "t1",
      toolName: "ask_questions",
      output: { status: "awaiting_user_response", questions: [] },
    } as unknown as Parameters<typeof renderPart>[0]);
  } finally {
    (process.stdout as { write: unknown }).write = original;
  }
  const out = written.join("");
  assert.match(out, /awaiting your answer/);
  assert.doesNotMatch(out, /error/);
});

test("startInstruction appends the idea, or nothing when there is none", () => {
  assert.equal(startInstruction("an expense tracker"), startInstructionText + ideaSteerPrefix + "an expense tracker");
  assert.equal(startInstruction(null), startInstructionText);
  assert.equal(startInstruction("   "), startInstructionText);
});
